<a href="https://colab.research.google.com/github/GalynaDe/ML/blob/main/HW_%D0%9C%D0%B5%D1%82%D0%BE%D0%B4%D0%B8_%D0%BF%D0%BE%D0%BD%D0%B8%D0%B6%D0%B5%D0%BD%D0%BD%D1%8F_%D1%80%D0%BE%D0%B7%D0%BC%D1%96%D1%80%D0%BD%D0%BE%D1%81%D1%82%D1%96.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Домашнє завдання: Пониження розмірностей для Аналізу Портретів Клієнтів

#### Контекст
В цьому ДЗ ми попрацюємо з методами пониження розмірності на наборі даних для задачі аналізу портретів клієнтів (Customer Personality Analysis). **В попередньому ДЗ ми працювали з цими даними використовуючи кластеризацію, зараз використаємо кластеризацію і візуалізауємо результати з різними методами.**

Customer Personality Analysis - це аналіз різних сегментів клієнтів компанії. Цей аналіз дозволяє бізнесу краще розуміти своїх клієнтів і полегшує процес адаптації продуктів під конкретні потреби, поведінку та інтереси різних типів клієнтів.

Аналіз портретів клієнтів допомагає бізнесу змінювати свій продукт на основі цільової аудиторії, розділеної на різні сегменти. Наприклад, замість того, щоб витрачати гроші на маркетинг нового продукту для всіх клієнтів у базі даних компанії, бізнес може проаналізувати, який сегмент клієнтів найімовірніше придбає продукт, і потім зосередити маркетингові зусилля лише на цьому сегменті.

#### Вхідні дані
Вам надано набір даних з такими атрибутами:

**Характеристики користувачів:**
- `ID`: Унікальний ідентифікатор клієнта
- `Year_Birth`: Рік народження клієнта
- `Education`: Рівень освіти клієнта
- `Marital_Status`: Сімейний стан клієнта
- `Income`: Річний дохід домогосподарства клієнта
- `Kidhome`: Кількість дітей у домогосподарстві клієнта
- `Teenhome`: Кількість підлітків у домогосподарстві клієнта
- `Dt_Customer`: Дата реєстрації клієнта у компанії
- `Recency`: Кількість днів з моменту останньої покупки клієнта
- `Complain`: 1, якщо клієнт скаржився за останні 2 роки, 0 - якщо ні

**Продукти:**
- `MntWines`: Сума, витрачена на вино за останні 2 роки
- `MntFruits`: Сума, витрачена на фрукти за останні 2 роки
- `MntMeatProducts`: Сума, витрачена на м'ясні продукти за останні 2 роки
- `MntFishProducts`: Сума, витрачена на рибні продукти за останні 2 роки
- `MntSweetProducts`: Сума, витрачена на солодощі за останні 2 роки
- `MntGoldProds`: Сума, витрачена на золото за останні 2 роки

**Акції:**
- `NumDealsPurchases`: Кількість покупок, зроблених з використанням знижок
- `AcceptedCmp1`: 1, якщо клієнт прийняв пропозицію у першій кампанії, 0 - якщо ні
- `AcceptedCmp2`: 1, якщо клієнт прийняв пропозицію у другій кампанії, 0 - якщо ні
- `AcceptedCmp3`: 1, якщо клієнт прийняв пропозицію у третій кампанії, 0 - якщо ні
- `AcceptedCmp4`: 1, якщо клієнт прийняв пропозицію у четвертій кампанії, 0 - якщо ні
- `AcceptedCmp5`: 1, якщо клієнт прийняв пропозицію у п'ятій кампанії, 0 - якщо ні
- `Response`: 1, якщо клієнт прийняв пропозицію в останній кампанії, 0 - якщо ні

**Взаємодія з компанією:**
- `NumWebPurchases`: Кількість покупок, зроблених через вебсайт компанії
- `NumCatalogPurchases`: Кількість покупок, зроблених за каталогом
- `NumStorePurchases`: Кількість покупок, зроблених безпосередньо у магазинах
- `NumWebVisitsMonth`: Кількість відвідувань вебсайту компанії за останній місяць


Для початку, запустіть код нижче. Всі ці кроки ми робили в попередньому ДЗ і для того, щоб результати кластеризації у нас були схожими, потрібно аби передобробка була однаковою.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd

# 1. Завантаження даних
df = pd.read_csv('drive/MyDrive/ML course 2024/HW 3.1/marketing_campaign.csv', sep='\t')

# 2. Обробка пропущених значень
df['Income_not_filled'] = df.Income.isna()
df.Income = df.Income.fillna(-1)

# 3. Обробка дати реєстрації
df.Dt_Customer = pd.to_datetime(df.Dt_Customer, format='%d-%m-%Y')
today = df.Dt_Customer.max()
df['days_lifetime'] = (today - df.Dt_Customer).dt.days
df['years_customer'] = df.Year_Birth.apply(lambda x: today.year - x)

# 4. Категоризація рівня освіти
df_education = pd.get_dummies(df.Education, prefix='education').astype(int)
df = pd.concat([df, df_education], axis=1)

# 5. Очищення сімейного стану
marital_status_map = {'Alone': 'Single', 'Absurd': 'Else', 'YOLO': 'Else'}
df['Marital_Status_clean'] = df.Marital_Status.map(marital_status_map)
df_ms = pd.get_dummies(df.Marital_Status_clean, prefix='marital').astype(int)
df = pd.concat([df, df_ms], axis=1)

# 6. Форматування доходу і видалення викиду
df.Income = df.Income.astype(int)
df = df[df.Income != 666666]

# 7. Створення фінального набору даних
X = df.drop(['ID', 'Dt_Customer', 'Education', 'Marital_Status', 'Marital_Status_clean'], axis=1)
X.reset_index(drop=True, inplace=True)

### Завдання 1: Виконання кластеризації та пониження розмірності для візуалізації результатів

Ваше завдання — провести кластеризацію клієнтів та візуалізувати результати кластеризації, використовуючи метод головних компонент (PCA) для пониження розмірності даних.

#### Інструкції:

1. **Вибір ключових характеристик:**
   Давайте обмежимось тільки наступними хараткеристиками для кластеризації цього разу:
   - `Income`: Річний дохід домогосподарства клієнта
   - `Recency`: Кількість днів з моменту останньої покупки клієнта
   - `NumStorePurchases`: Кількість покупок, зроблених безпосередньо у магазинах
   - `NumDealsPurchases`: Кількість покупок, зроблених з використанням знижок
   - `days_lifetime`: Кількість днів з моменту реєстрації клієнта у компанії
   - `years_customer`: Вік клієнта
   - `NumWebVisitsMonth`: Кількість відвідувань вебсайту компанії за останній місяць
   Відберіть в наборі даних `X` лише ці характеристики.

2. **Нормалізація даних:**
   Використайте метод `MinMaxScaler` для нормалізації значень обраних характеристик.

3. **Кластеризація:**
   Проведіть кластеризацію клієнтів, використовуючи метод `KMeans` з трьома кластерами.

4. **Пониження розмірності:**
   Використайте метод головних компонент (PCA) для пониження розмірності даних до трьох компонент.

5. **Візуалізація результатів:**
   Використовуючи plolty express побудуйте 3D-графік розподілу клієнтів у просторі трьох головних компонент, де кольором позначено кластери.

6. **Опишіть, що спостерігаєте:**
   Чи кластеризація чітко розділила дані?

Далі ми детальніше проінтерпретуємо результати візуалізації і пониження розмірностей.

In [4]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
import plotly.express as px
import pandas as pd  # Убедитесь, что библиотека pandas импортирована

# Вибір ключових характеристик
selected_columns = ['Income', 'Recency', 'NumStorePurchases', 'NumDealsPurchases',
                    'days_lifetime', 'years_customer', 'NumWebVisitsMonth']
X = X[selected_columns]

# Нормалізація даних
scaled_X = pd.DataFrame(MinMaxScaler().fit_transform(X), columns=selected_columns)

# Кластеризація
clustering = KMeans(n_clusters=3, n_init='auto')
clustering.fit(scaled_X)

# Прогнозування кластерів
labels = clustering.predict(scaled_X)

# Додавання міток кластерів до даних
cluster_df = X.copy()
cluster_df['labels'] = labels

# Пониження розмірності за допомогою PCA
pca = PCA(n_components=3, random_state=0)
pca_df = pd.DataFrame(pca.fit_transform(cluster_df.drop('labels', axis=1)), columns=['PC1', 'PC2', 'PC3'])
pca_df['labels'] = cluster_df.labels

# Візуалізація результатів у 3D
fig = px.scatter_3d(
    pca_df,
    x='PC1',
    y='PC2',
    z='PC3',
    color='labels',
    title='3D Scatter Plot of Dataset with PCA'
)

fig.show()


### Завдання 2: Аналіз результатів пониження розмірності

1. **Розрахунок частки поясненої дисперсії:**
   Визначте, яка частка загальної варіації даних пояснюється кожною з трьох головних компонент (PC1, PC2, PC3) за допомогою атрибуту `explained_variance_ratio_` об'єкта PCA. Виведіть результат на екран.

2. **Розрахунок кумулятивної частки поясненої дисперсії:**
   Обчисліть кумулятивну частку поясненої дисперсії для трьох головних компонент, щоб зрозуміти, скільки варіації даних пояснюється першими кількома компонентами.

In [5]:
import numpy as np

# Розрахунок частки поясненої дисперсії
explained_variance_ratio = pca.explained_variance_ratio_

# Розрахунок кумулятивної частки поясненої дисперсії
cumulative_variance_ratio = np.cumsum(explained_variance_ratio)

# Результати
print("Частка поясненої дисперсії для кожної компоненти:")
for i, var_ratio in enumerate(explained_variance_ratio):
    print(f"PC{i+1}: {var_ratio:.2%}")

print("\nКумулятивна частка поясненої дисперсії:")
for i, cum_var in enumerate(cumulative_variance_ratio):
    print(f"PC1 to PC{i+1}: {cum_var:.2%}")

Частка поясненої дисперсії для кожної компоненти:
PC1: 99.99%
PC2: 0.01%
PC3: 0.00%

Кумулятивна частка поясненої дисперсії:
PC1 to PC1: 99.99%
PC1 to PC2: 100.00%
PC1 to PC3: 100.00%


Перша компонента описує 99.99% варіативності даних.

### Завдання 3: Інтерпретація "Loadings"

Продовжуємо інтерпретацію результатів `PCA`і познайомимось з новим поняттям `loadings`, яке допоможе нам знайти звʼязок між головними компонентами і оригінальними ознаками в наборі даних.

Ми зараз побудували візуалізацію кластерів точок даних в просторі трьох головних компонент. Але хочеться знайти звʼязок між головними компонентами і оригінальними ознаками. Для розуміння, які початкові характеристики даних мають найбільший вплив на ці головні компоненти, ми можемо використати атрибут `components_` методу `PCA`.

#### Що таке `pca.components_`?

`pca.components_` — це масив, який містить коефіцієнти (або "ваги"), що показують внесок кожної вихідної ознаки у кожну з головних компонент. Ці коефіцієнти ще називаються **"loading"** або "навантаженнями" компонент.

- **Loadings** (`навантаження`) відображають важливість кожної змінної (ознаки) для відповідної головної компоненти. Вони показують, яким чином змінні поєднуються, щоб утворити нові, зменшені вимірювання.
- Якщо коефіцієнт має високе абсолютне значення (як позитивне, так і негативне), це вказує на те, що відповідна змінна сильно впливає на головну компоненту.

#### Саме завдання
Ваше завдання — обчислити "навантаження" для кожної з головних компонент і інтерпретувати результати.

1. **Обчислення loadings для компонент:**
   Використайте атрибут `components_` об'єкта PCA для створення DataFrame, який відображатиме внесок кожної вихідної ознаки в кожну головну компоненту.

2. **Інтерпретація результатів:**
   Виведіть значення "навантажень" і проаналізуйте, які ознаки найбільше впливають на кожну головну компоненту.

In [7]:
# Обчислення "навантажень" для компонент
loadings = pd.DataFrame(pca.components_.T, columns=['PC1', 'PC2', 'PC3'], index=cluster_df.drop('labels', axis=1).columns)

print("\nPCA Loadings:")
loadings.round(2)


PCA Loadings:


PC1  PC2   PC3
Income             1.0  0.0 -0.00
Recency           -0.0  0.0  1.00
NumStorePurchases  0.0  0.0 -0.00
NumDealsPurchases -0.0  0.0 -0.00
days_lifetime     -0.0  1.0 -0.00
years_customer     0.0 -0.0  0.01
NumWebVisitsMonth -0.0  0.0 -0.00

###Завдання 4
Давайте проаналізуємо "навантаження" (**loadings**) для трьох головних компонент після вилучення ознаки `Income`. Це допоможе нам зрозуміти, як змінилася важливість інших ознак для кожної головної компоненти, коли одна з ключових характеристик (`Income`) була вилучена.

#### Кроки для проведення аналізу і ваше завдання:

1. Видаліть ознаку `Income` з нашого набору даних `X` і повторно виконайте PCA (метод головних компонент) для отримання нових "навантажень".

2. Обчисліть нові "навантаження" для трьох головних компонент на наборі даних без `Income`

3. Проаналізуйте, які ознаки мають найбільший вплив на кожну головну компоненту після вилучення `Income`.

4. Перегляньте, наскільки кожна з головних компонент пояснює дисперсію в даних без ознаки `Income`.

In [8]:
# Вилучення ознаки 'Income' з даних
selected_columns_no_income = ['Recency', 'NumStorePurchases', 'NumDealsPurchases',
                    'days_lifetime', 'years_customer', 'NumWebVisitsMonth']

X_without_income = cluster_df[selected_columns_no_income]

# Виконання PCA без ознаки 'Income'
pca = PCA(n_components=3, random_state=0)
pca.fit(X_without_income)

# Обчислення "навантажень" для компонент
loadings_without_income = pd.DataFrame(pca.components_.T, columns=['PC1', 'PC2', 'PC3'], index=X_without_income.columns)

# Виведення результатів
print("\nPCA Loadings без 'Income':")
loadings_without_income.round(2)


PCA Loadings без 'Income':


PC1   PC2   PC3
Recency            0.0  1.00 -0.01
NumStorePurchases  0.0 -0.00  0.04
NumDealsPurchases  0.0 -0.00  0.01
days_lifetime      1.0 -0.00  0.00
years_customer    -0.0  0.01  1.00
NumWebVisitsMonth  0.0 -0.00 -0.02

Видно, що days_lifetime пояснює найбільшу частину дисперсії, отже, це друга за важливістю характеристика для класифікації клієнтів.

### Завдання 5: Візуалізація кластеризації за допомогою t-SNE

Ваше завдання — використати метод t-SNE для візуалізації результатів кластеризації клієнтів у двовимірному просторі. Метод t-SNE допомагає знизити розмірність даних та зберегти локальні структури в даних, що робить його ефективним для візуалізації високорозмірних даних. Ми також зможемо порівняти результат цього методу з РСА.

1. Використайте метод t-SNE для зниження розмірності до 2х вимірів даних, які включають ознаки всі, що і в завданні 1, а також були відмасштабовані перед пониженням розмірностей.

2. Створіть новий DataFrame з координатами, отриманими після застосування t-SNE, та додайте до нього мітки кластерів.

3. Побудуйте інтерактивний 2D-графік розподілу клієнтів, де кольором буде позначено різні кластери і проаналізуйте графік з рекомендаціями нижче (можливо треба буде вивести додаткові візуалізації чи таблиці для інтерпретації, але треба прям зрозуміти, які ознаки формують який кластер і чим кластери відрізняються одне від одного).

  **Опишіть отримані кластери з точки зору ознак.**

4. Опишіть відмінність графіка tSNE від PCA.

#### ЯК можна інтерпретувати з t-SNE?

Хоча t-SNE не надає "компонентів" як РСА, він забезпечує низьковимірне представлення даних, яке можна візуально інтерпретувати:

- **Кластери:** t-SNE особливо добре показує кластери подібних точок. Якщо ви бачите чітко визначені кластери на графіку t-SNE, це свідчить про наявність груп схожих спостережень у ваших даних. Проаналізувати їх можемо, якщо додамо дані в `hover_data` або якщо якісь з даних виведемо як розмір чи форма точок на візуалізації. Також корисно може бути вивести середні значення ознак по кластерам.
- **Локальна структура:** Відносне розташування точок одного кластеру на графіку t-SNE може допомогти вам зрозуміти, які дані подібні між собою.
- **Глобальна структура:** Будьте обережні; t-SNE менш надійний для відображення глобальних структур (наприклад, відстаней між кластерами) у порівнянні з PCA, бо t-SNE націлений на збереження саме локальних структур.

In [9]:
from sklearn.manifold import TSNE

# Пониження розмірності за допомогою t-SNE
tsne = TSNE(n_components=2, random_state=0)
tsne_df = pd.DataFrame(tsne.fit_transform(scaled_X), columns=['x', 'y'])

# Додавання міток кластерів
tsne_df['labels'] = labels
tsne_df = pd.concat([tsne_df, X], axis=1)

fig = px.scatter(
    tsne_df,
    x='x',
    y='y',
    hover_data=['labels'] + selected_columns,
    color='labels',
    title='2D Scatter Plot of Dataset with t-SNE'
)
fig.show()

In [12]:
tsne_df.groupby('labels')[selected_columns].mean().style.background_gradient(cmap='Oranges')

Опис кластерів:

Кластер 0: Ця група клієнтів характеризується високим рівнем доходу. Вони переважно здійснюють покупки у фізичних магазинах і рідко користуються знижками. Вебсайт вони відвідують менш активно, порівняно з іншими кластерами. Це може свідчити про їхню орієнтацію на традиційний шопінг, а не на онлайн-торгівлю.

Кластер 1: Учасники цього кластеру мають найнижчий рівень доходу, активно шукають знижки та частіше взаємодіють із вебсайтом компанії. Вони довго співпрацюють із компанією, маючи найбільшу тривалість співпраці (lifetime) і короткий проміжок часу з останньої покупки (Recency). Це може вказувати на те, що вони є клієнтами з обмеженим бюджетом, які цінують вигідні пропозиції та онлайн-можливості.

Кластер 2: Клієнти цього кластеру мають середній дохід, рідко користуються знижками й помірно взаємодіють із вебсайтом. Вони належать до нової клієнтської бази, оскільки нещодавно почали співпрацювати з компанією і ще не сформували сталі звички покупок.
